### ISSUE TITLE
Unnable to connect to logentries from a Linux server

### ISSUE THREAD

**Issue by azumelzu** [zyborg/LogentriesDotNet#5]
@ebekker @aleksandrov 

Hey Guys, 
Anytime I try to connect to logentries from linux I get the following error:

LE: Unable to connect to Logentries API.
System.IO.IOException: An error occurred while opening the connection. ---> System.PlatformNotSupportedException: Operation is not supported on this platform.
   at System.Net.Sockets.SocketPal.WindowsIoctl(SafeCloseSocket handle, Int32 ioControlCode, Byte[] optionInValue, Byte[] optionOutValue, Int32& optionLength)
   at System.Net.Sockets.Socket.IOControl(Int32 ioControlCode, Byte[] optionInValue, Byte[] optionOutValue)
   at Zyborg.Logentries.LogentriesClient.SetSocketKeepAliveValues(TcpClient tcpc, Int32 KeepAliveTime, Int32 KeepAliveInterval) in /var/lib/jenkins/jobs/core-basic-backend-int/workspace/src$
   at Zyborg.Logentries.LogentriesClient.Connect() in /var/lib/jenkins/jobs/core-basic-backend-int/workspace/src/Zyborg.Logentries/LogentriesClient.cs:line 118
   at Zyborg.Logentries.LogentriesAsyncLogger.OpenConnection() in /var/lib/jenkins/jobs/core-basic-backend-int/workspace/src/Zyborg.Logentries/LogentriesAsyncLogger.cs:line 340
   --- End of inner exception stack trace ---
   at Zyborg.Logentries.LogentriesAsyncLogger.OpenConnection() in /var/lib/jenkins/jobs/core-basic-backend-int/workspace/src/Zyborg.Logentries/LogentriesAsyncLogger.cs:line 352
   at Zyborg.Logentries.LogentriesAsyncLogger.ReopenConnection() in /var/lib/jenkins/jobs/core-basic-backend-int/workspace/src/Zyborg.Logentries/LogentriesAsyncLogger.cs:line 366

I also tried to resolve the ip address manually but it didn't work.

It works ok from Windows, using the default url (api.logentries.com) or any ip returned by that host:

I resolved the ip address modifying a little bit the Connect metgod on LogentriesClient:

 public void Connect()
        {
            m_Client = new TcpClient();

            var addressEndpoint = new DnsEndPoint(LeApiUrl, 10000);
            var hostEntry = Dns.GetHostEntryAsync(addressEndpoint.Host).Result;
            m_Client.ConnectAsync(hostEntry.AddressList.First(x => x.ToString() == "54.217.226.19"), m_TcpPort).Wait();
            m_Client.NoDelay = true;

            // on Azure sockets will be closed after some minutes idle.
            // which for some reason messes up this library causing it to lose data.
            
            // turn on keepalive, to keep the sockets open.
            // I don't really understand why this helps the problem, since the socket already has NoDelay set
            // so data should be sent immediately. And indeed it does appear to be sent promptly when it works.
            m_Client.Client.SetSocketOption( SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);

            // set timeouts to 10 seconds idle before keepalive, 1 second between repeats, 
            SetSocketKeepAliveValues(m_Client, 10 *1000, 1000);
            
            m_Stream = m_Client.GetStream();

            Console.WriteLine($"Logentries -- stream can read: { m_Stream.CanRead }");

            if (m_UseSsl)
            {
                m_SslStream = new SslStream(m_Stream);
                m_SslStream.AuthenticateAsClientAsync(m_ServerAddr).Wait();
            }
        }

This is running on AWS - Ubuntu 14.04

Did you experience any similar issues? 

Thanks,
Andres

**Comment by azumelzu** [comment]
@ebekker @aleksandrov 

I fixed the issue. 

The problem is that socket.IOControl is not supported in all platforms by now (is not supported on Linux)

The fix is just to comment out the line that calls SetSocketKeepAliveValues in LogentriesClient which internally calls tcpc.Client.IOControl(IOControlCode.KeepAliveValues, inOptionValues, null);

Doing that, I'm able to send data to Logentries from Ubuntu.

Please, correct me if I'm wrong, but that was an attempt to fix an issue with Azure, right?

Thanks,
Andres

**Comment by ebekker** [comment]
That part of the code was just pulled as-is from the original .NET code base, so I'm not sure what was the purpose of the code.  If it's simply not supported on Linux, we can either do a conditional compile-time guard (pre-proc directives) to exclude it for Linux or do a run-time guard by testing against the runtime environment.

I'm not actively working on this tool right now, so since @aleksandrov already submitted a PR to address this (I missed it when he submitted it originally) -- I'm just going to pull it in, and it would be great if @azumelzu  could test and confirm the fix.

**Comment by ebekker** [comment]
PR #2 has been merged, if you can confirm it works we can close out this issue.

**Comment by azumelzu** [comment]
@ebekker @aleksandrov I'm going to give it a try and I'll get back to you.

Is this being updated on nuget?

Thanks,

**Comment by azumelzu** [comment]
@ebekker @aleksandrov 

Could you please add this little check around  SetSocketKeepAliveValues? That way we can avoid the exception on every connect call.
```csharp
try
{
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
    {
        SetSocketKeepAliveValues(m_Client, 10 * 1000, 1000);
    }
}
catch (PlatformNotSupportedException)
{
}
```

Thanks!

**Comment by ebekker** [comment]
Done

**Comment by ghost** [comment]
@ebekker 

Can this be pushed to NuGet? We are currently experiencing difficulties on Linux as well (no log entries) (windows all works). Would prefer to use the nuget feed version.

Thank you in advance

**Comment by ebekker** [comment]
Published as `0.6.0`.
